### ISSUE TITLE
Blockly#getMainWorkspace is unreliable (Customizable block colors causes crash when making custom block)

### ISSUE THREAD

**Issue by callumjt** [ScratchAddons/ScratchAddons#6212]
### Description

when you click on "create block" the editor crashes and you have to reload
![image](https://github.com/ScratchAddons/ScratchAddons/assets/116295240/f165f566-2ccc-403b-b0ab-6e18a4ae7a6c)

### Steps to reproduce

1.turn on Customizable block colors
2.try to create a custom block

### Environment

Browser: firefox 114.0.1
Scratch Addons version: 1.33
Operating system: windows 11

**Comment by DNin01** [comment]
Can reproduce.

**Comment by medlans** [comment]
Can replicate.
Browser: Chrome 114.0
Scratch Addons version: 1.33 (it didn't happen before this version)
Operating system: MacOS X Macintosh 11.3
![flash](https://github.com/ScratchAddons/ScratchAddons/assets/124630024/dca58d8e-b7c8-434d-b922-4731650590f1)
I turned that addon off and was able to create one:
![off](https://github.com/ScratchAddons/ScratchAddons/assets/124630024/c7a8a544-80ff-467d-8591-0611bcc79f6e)

**Comment by Samq64** [comment]
This looks like an `editor-theme3` crash.
```
gui Unhandled Error with action='Blocks': TypeError: Cannot read properties of null (reading 'getWorkspace')    at updateColors (chrome-extension://ahjndmkmpohijgkaglnfmmiiflfpkcll/addons/editor-theme3/theme3.js:439:36) ...
```

**Comment by apple502j** [comment]
This affects most, if not all, Blockly addons.

**Comment by apple502j** [comment]
This basically breaks all addons that use `Blockly#getMainWorkspace` and `Blockly#mainWorkspace`. After a definition prompt is opened, it returns the prompt's detached workspace, instead of the actual workspace. A workspace refresh, triggered by going to and back from project page, resets it to actual one.

This bug has existed since the very early versions - but it only became apparent with this crash.

The solution is to query the workspace not from `Blockly`, but from the React state where we grab the `ScratchBlocks` from. However, since multiple instances may exist during an addon's life, we cannot store this in a const, instead querying on every occasion. While I believe the performance cost is low, not a easy patch when this affects dozens of addons.

@towerofnix @GarboMuffin  Is my analysis correct?

**Comment by GarboMuffin** [comment]
Yes, that matches some prior observations I've had

I haven't looked, but I would expect there is way to get from the custom block editing workspace to the real main workspace. Seems like something that could go on addon.tab

**Comment by WorldLanguages** [comment]
Is this a regression after a Scratch update? Or because of our 1.33.0 update? 🤔

**Comment by medlans** [comment]
> Is this a regression after a Scratch update? Or because of our 1.33.0 update? 🤔

I tested it in 1.32.1, and it was fine, but when I updated to 1.33.0, it started doing this.

**Comment by Hans5958** [comment]
Issue pinned and renamed for relevance and visibility.

**Comment by WorldLanguages** [comment]
It appears that #6216 would revert the commit that caused 1.33.0 to regress, even though the root cause of the bug predates this version by many months apparently.

**Comment by towerofnix** [comment]
@apple502j I'm only experienced with Blockly internals (as used in scratch-blocks), not with external interfaces and handling multiple extant workspaces at once (as in scratch-gui and the procedure creation component), so I can't peer-review your analysis as closely as I'd like to. But it seems about right to me. Pulling from React sounds like the most appropriate way to go, albeit potentially inconvenient to refactor across multiple addons, yeah.

This is absolutely not a detail that addons should be concerning themselves with, so is there a function we can conveniently change the internals of today? Or do addons just all use `Blockly.getMainWorkspace` today? I'd think they should be changed to a utility available to all addons which is implementation-agnostic, i.e. `getMainEditorWorkspace()` or whatever, and then we can implement the correct internals now and change it later if needed without having to refactor addons again.

Side-note: This would totally have slipped by me too LOL, but I have to recognize that [we were warned](https://developers.google.com/blockly/reference/js/blockly.getmainworkspace_variable)...

> `Blockly.getMainWorkspace`: Returns the main workspace. Returns the last used main workspace (based on focus). **Try not to use this function, particularly if there are multiple Blockly instances on a page.**

**Comment by kokofixcomputers** [comment]
Can reproduce.

**Comment by WorldLanguages** [comment]
The reproduction described in the first post was fixed by #6216, which simply reverted a recent change to editor-theme3. The underlying root cause of the issue discussed above this comment isn't fixed yet.

@kokofixcomputers Bug should be fixed now, in the master branch. Ping me if that's not the case.

**Comment by TheElementGuy** [comment]
Can reproduce on Edge. The computer is Windows.

**Comment by medlans** [comment]
> Can reproduce on Edge. The computer is Windows.

The issue has been fixed in v1.33.1.

**Comment by WorldLanguages** [comment]
Related: most crashes can be avoided if we wrap our own code in `try{/*our-code*/} catch{}` when overriding existing VM/Blockly methods.

**Comment by WorldLanguages** [comment]
We'll be using this issue to track that `Blockly.getMainWorkspace` is unreliable. See this comment for more info: https://github.com/ScratchAddons/ScratchAddons/pull/6227#issuecomment-1599952073

The crash is related to what I just mentioned, but is already fixed on master branch.

**Comment by WorldLanguages** [comment]
PR #6613 will add `addon.tab.traps.getWorkspace()`, but so far it will only be used to fix bug #3906
